scylla-macros/src/serialize/row.rs

```diff
@@ -195,94 +203,130 @@ struct ColumnSortingGenerator<'a> {
+        let struct_name = &self.ctx.struct_name;
+        let (impl_generics, ty_generics, where_clause) = self.ctx.generics.split_for_impl();
+        let partial_struct_name = syn::Ident::new(
+            &format!("_{}ScyllaSerPartial", struct_name),
+            struct_name.span(),
+        );
+        let mut partial_generics = self.ctx.generics.clone();
+        let partial_lt: syn::LifetimeParam = syn::parse_quote!('scylla_ser_partial);
+        if !self.ctx.fields.is_empty() {
+            partial_generics
+                .params
+                .push(syn::GenericParam::Lifetime(partial_lt.clone()));
+        }
 
-        let rust_field_idents = self
-            .ctx
-            .fields
-            .iter()
-            .map(|f| f.ident.clone())
-            .collect::<Vec<_>>();
-        let rust_field_names = self
-            .ctx
-            .fields
-            .iter()
-            .map(|f| f.column_name())
-            .collect::<Vec<_>>();
-        let udt_field_names = rust_field_names.clone(); // For now, it's the same
+        let (partial_impl_generics, partial_ty_generics, partial_where_clause) =
+            partial_generics.split_for_impl();
 
-        // Declare a helper lambda for creating errors
-        statements.push(self.ctx.generate_mk_typck_err());
-        statements.push(self.ctx.generate_mk_ser_err());
-
-        // Generate a "visited" flag for each field
-        let visited_flag_names = rust_field_idents
-            .iter()
-            .map(|s| syn::Ident::new(&format!("visited_flag_{}", s), Span::call_site()))
-            .collect::<Vec<_>>();
-        statements.extend::<Vec<_>>(parse_quote! {
-            #(let mut #visited_flag_names = false;)*
-        });
+        let columns: Vec<_> = self.ctx.fields.iter().map(|f| f.column_name()).collect();
+        let fields: Vec<_> = self.ctx.fields.iter().map(|f| &f.ident).collect();
+        let tys: Vec<_> = self.ctx.fields.iter().map(|f| &f.ty).collect();
 
-        // Generate a variable that counts down visited fields.
-        let field_count = self.ctx.fields.len();
-        statements.push(parse_quote! {
-            let mut remaining_count = #field_count;
-        });
+        let all_names = self.ctx.fields.iter().map(|f| f.column_name());
 
-        // Generate a loop over the fields and a `match` block to match on
-        // the field name.
-        statements.push(parse_quote! {
-            for spec in ctx.columns() {
+        let partial_struct: syn::ItemStruct = parse_quote! {
+            pub struct #partial_struct_name #partial_generics {
+                #(#fields: &#partial_lt #tys,)*
+                missing: ::std::collections::HashSet<&'static str>,
+            }
+        };
+
+        let serialize_field_block: syn::Block = if self.ctx.fields.is_empty() {
+            parse_quote! {{
+                ::std::result::Result::Ok(#crate_path::ser::row::FieldStatus::NotUsed)
+            }}
+        } else {
```

Comment:
When I said to not worry about the hashset I was not aware that it would also be necessary for the case where `flatten` is not used. This introduces new overhead to the case that does not need it.

Before we used local variables (one per field) to track completion. The obvious way to migrate that to partial struct would be to have bool fields in the struct, instead of local variables. Why do we need a set instead of that?

Reply:
When I first wrote this it got into my head that I *really* needed the hashset because I couldn't keep track of the flattened fields at compile time but I think that may have been the case for a previous implementation that ended up not going anywhere. I think the current implementation could probably just be done with one boolean per field in the struct, I'll give it a shot.

Reply:
Done. I have it now the same way as it used to be: a counter that keeps track of how many fields are remaining to serialize, and a per-field boolean to check what fields have already been visited.